_posts/2019-01-03-elixir-kaffe-codealong.md

```diff
@@ -0,0 +1,204 @@
+
+### Set up Kafka Server
+Follow the first two steps of the [quickstart instructions](http://kafka.apache.org/documentation/#quickstart) from Apache Kafka:
+1. [Download the code](https://www.apache.org/dyn/closer.cgi?path=/kafka/2.1.0/kafka_2.11-2.1.0.tgz)
+2. Start the servers
+Zookeeper
+`bin/zookeeper-server-start.sh config/zookeeper.properties`
+Kafka
+`bin/kafka-server-start.sh config/server.properties`
+
+### Set up Elixir App
+
+**1. Start new project**
+`mix new elixir_kaffe_codealong`
+
+**2. Configure kaffe**
+**2.a:** In `mix.exs` add `:kaffe` to the list of extra applications:
+```elixir
+def application do
+  [
+    extra_applications: [:logger, :kaffe]
+  ]
+end
+```
+
+**2.b:** Add kaffe to list of dependencies:
+```elixir
+defp deps do
+  [
+    {:kaffe, "~> 1.9"}
+  ]
+end  
+```
+
+**2.c:** Run `mix deps.get` in the terminal to lock new dependencies.
+
+**3. Configure producer**
+in `config/config.exs` add:
+```elixir
+config :kaffe,
+  producer: [
+    endpoints: [localhost: 9092],
+    # endpoints references [hostname: port]. Kafka is configured to run on port 9092.
+    # In this example, the hostname is localhost because we've started the Kafka server
+    # straight from our machine. However, if the server is dockerized, the hostname will
+    # be called whatever is specified by that container (usually "kafka")
+    topics: ["our_topic", "another_topic"], # add a list of topics you plan to produce messages to
+  ]
+```
+
+**4. Configure consumer**
+**4.a:** add `/lib/application.ex` with the following code:
+```elixir
+defmodule ElixirKaffeCodealong.Application do
+  use Application # read more about Elixir's Application module here: https://hexdocs.pm/elixir/Application.html
+
+  def start(_type, args) do
+    import Supervisor.Spec
+    children = [
+      worker(Kaffe.Consumer, []) # calls to start Kaffe's Consumer module
+    ]
+    opts = [strategy: :one_for_one, name: ExampleConsumer.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+end
+```
+**4.b:** back in `mix.exs`, add a new item to the application function:
+```elixir
+def application do
+  [
+    extra_applications: [:logger, :kaffe],
+    mod: {ElixirKaffeCodealong.Application, []}
+    # now that we're using the Application module, this is where we'll tell it to start.
+    # We use the keyword `mod` with applications that start a supervision tree,
+    # which we configured when adding our Kaffe.Consumer to Application above.
+  ]
+end
+```
+**4.c:** add a consumer module to accept messages from Kafka as `/lib/example_consumer.ex` with the following code:
+```elixir
+defmodule ExampleConsumer do
+  # function to accept Kafka messaged MUST be named "handle_message"
+  # MUST accept arugments structured as shown here
+  # MUST return :ok
+  # Can do anything else within the function with the incoming message
+
+  def handle_message(%{key: key, value: value} = message) do # necessary
+    IO.inspect message # optional
```

Comment:
Let's remove the `# optional` comments, folks reading this will know `IO.*` calls are as such 👍 

Additionally can we use parenthesis for functions?

```elixir
IO.inspect(message)
```